MapboxGeocoder/MBPlacemark.swift

```diff
@@ -62,44 +63,65 @@ public let MBPostalAddressISOCountryCodeKey = "ISOCountryCode"
  */
 @objc(MBPlacemark)
-open class Placemark: NSObject, NSSecureCoding {
-    /**
-     The GeoJSON feature dictionary containing the placemark’s data.
-     */
-    fileprivate let featureJSON: JSONDictionary
+open class Placemark: NSObject, Codable {
+    
+    private enum CodingKeys: String, CodingKey {
+        case identifier = "id"
+        case name = "text"
+        case address
+        case qualifiedName = "place_name"
+        case superiorPlacemarks = "context"
+        case centerCoordinate = "center"
+        case code = "short_code"
+        case wikidataItemIdentifier = "wikidata"
+        case properties
+        case boundingBox = "bbox"
+    }
     
     /**
      Creates a placemark from the given [Carmen GeoJSON](https://github.com/mapbox/carmen/blob/master/carmen-geojson.md) feature.
      */
-    internal init(featureJSON: JSONDictionary) {
-        self.featureJSON = featureJSON
-    }
-    
-    @objc public convenience required init?(coder aDecoder: NSCoder) {
-        guard let featureJSON = aDecoder.decodeObject(of: NSDictionary.self, forKey: "featureJSON") as? JSONDictionary else {
-            return nil
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        identifier = try container.decode(String.self, forKey: .identifier)
+        name = try container.decode(String.self, forKey: .name)
+        address = try container.decodeIfPresent(String.self, forKey: .address)
+        qualifiedName = try container.decodeIfPresent(String.self, forKey: .qualifiedName)
+        superiorPlacemarks = try container.decodeIfPresent([GeocodedPlacemark].self, forKey: .superiorPlacemarks)
+        
+        if let coordinates = try container.decodeIfPresent([CLLocationDegrees].self, forKey: .centerCoordinate) {
+            let coordinate = CLLocationCoordinate2D(geoJSON: coordinates)
+            location = CLLocation(coordinate: coordinate)
         }
         
-        self.init(featureJSON: featureJSON)
-    }
-    
-    /**
-     Creates a placemark with the same data as another placemark object.
-     */
-    @objc public convenience init(placemark: Placemark) {
-        self.init(featureJSON: placemark.featureJSON)
-    }
-    
-    @objc public class var supportsSecureCoding : Bool {
-        return true
-    }
-    
-    @objc open func copy(with zone: NSZone?) -> AnyObject {
-        return Placemark(featureJSON: featureJSON)
+        code = try container.decodeIfPresent(String.self, forKey: .code)?.uppercased()
+        if let identifier = try container.decodeIfPresent(String.self, forKey: .wikidataItemIdentifier) {
+            assert(identifier.hasPrefix("Q"))
+            wikidataItemIdentifier = identifier
```

Comment:
Looks like we’re optimistically decoding the QualifyingPlacemark properties in the superclass’s decoder. That’s alright, since GeocodedPlacemark and QualifyingPlacemark don’t have any identically named properties with different implementations. However, it probably points to a need to eventually turn Placemark itself into a protocol.